docs/language-server/DocumentEventSyncPublisherSubscriberDesign.md

```diff
@@ -0,0 +1,114 @@
+# Document event sync publisher subscriber
+
+## Overview
+
+This document contains information about the implementation of a notification system inside the language server using the 
+PubSub architecture. Goal is to allow communication between language server and extensions synchronously.
+
+Following diagram shows how this has been implemented. 
+
+![alt text](../images/language-server/001-DocumentEventPubSub-class-diagram.png)
+
+- Event Subscribers and Event Publishers get loaded using JavaSPI and get subscribed to publishers when a document 
+opens or changes.
+- Event Publishers and Subscribers have an Event kind and a subscriber can only subscribe to a single publisher at once.
+If we need a subscriber to subscribe for two publishers, then we have to create two different subscribers and subscribe
+separately.
+- When an Event Publisher publishes, the `onEvent()` method is getting triggered for each subscriber that has been
+subscribed to the publisher.
+- We can override `publish()` and `onEvent()` methods to control the event synchronization.
+  
+
+## How to Use
+
+- **Event Publisher** is an internal interface and cannot be accessed publicly by extension developers. The server 
+decides which events to be delegated to a particular publisher. In the `publish()` method a debouncing mechanism has 
+implemented for the **Project Update Event Publisher**.
```

Comment:
Classes we should use `ClassName` formatting. Not bold font. Same for the `EventPublisher` and `EventSubscriber` interfaces.